### PR TITLE
Bump hiredis to v1.2.0

### LINKED ISSUE #40279: [Core] Shepherding Security vulerability issue from hiredis dependency
### What happened + What you expected to happen

https://github.com/ray-project/ray/pull/40217

cc @thomasdesr any details you can add? 

### Versions / Dependencies

master

### Reproduction script

n/a

### Issue Severity

None

### PR DESCRIPTION
**Pull request by thomasdesr**
## Why are these changes needed?

Bump to latest to fix outstanding vulnerabilities

## Related issue number

Closes https://github.com/ray-project/ray/issues/40279

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Hmm actually all the windows build seem to fail?

File: bazel/ray_deps_setup.bzl [hunk 1]
```diff
@@ -118,8 +118,8 @@ def ray_deps_setup():
     auto_http_archive(
         name = "com_github_redis_hiredis",
         build_file = "@com_github_ray_project_ray//bazel:BUILD.hiredis",
-        url = "https://github.com/redis/hiredis/archive/392de5d7f97353485df1237872cb682842e8d83f.tar.gz",
-        sha256 = "2101650d39a8f13293f263e9da242d2c6dee0cda08d343b2939ffe3d95cf3b8b",
+        url = "https://github.com/redis/hiredis/archive/60e5075d4ac77424809f855ba3e398df7aacefe8.tar.gz",
+        sha256 = "b6d6f799b7714d85316f9ebfb76a35a78744f42ea3b6774289d882d13a2f0383",
         patches = [
             "@com_github_ray_project_ray//thirdparty/patches:hiredis-windows-msvc.patch",
```
Comment (mattip) [inline bazel/ray_deps_setup.bzl:124]: Maybe try deleting this patch, or at least the part about strdup?

Comment (rkooo567) [reply bazel/ray_deps_setup.bzl:124]: Upgrade is necessary for the vulnerability fix

Comment (mattip) [reply bazel/ray_deps_setup.bzl:124]: OK

Comment (thomasdesr) [reply bazel/ray_deps_setup.bzl:124]: Thanks @mattip!

**Comment by mattip** [comment]
This diff will fix the failure in `alloc.c`. There is still a failure to build boringssl. I don't really understand a few things:
- what changed, why does master build boringssl and this version does not?
- why use [boringssl for hiredis](https://github.com/ray-project/ray/blob/65ed62d512720c037f545d52cfc653dc830a4589/bazel/BUILD.hiredis#L48), and [openssl for redis](https://github.com/ray-project/ray/blob/65ed62d512720c037f545d52cfc653dc830a4589/bazel/BUILD.redis#L35)? It seems openssl is more prevalent, especially in the documentation.
- If we are talking about security, why is openssl [pinned to the very old openssl v1.1.1f](https://github.com/ray-project/ray/blob/65ed62d512720c037f545d52cfc653dc830a4589/bazel/ray_deps_setup.bzl#L252) when there are security releases, the final 1.1.1 version [is 1.1.1w](https://github.com/openssl/openssl/tags)

Anyhow, here is what I have so far
```diff
diff --git a/thirdparty/patches/hiredis-windows-msvc.patch b/thirdparty/patches/hiredis-windows-msvc.patch
index 2d05b117c4..733cb75665 100644
--- a/thirdparty/patches/hiredis-windows-msvc.patch
+++ b/thirdparty/patches/hiredis-windows-msvc.patch
@@ -17,7 +17,7 @@ diff --git sds.h sds.h
 diff --git fmacros.h fmacros.h
 --- fmacros.h
 +++ fmacros.h
-@@ -12,0 +12,3 @@
+@@ -13,0 +13,3 @@
 +#if defined(_MSC_VER) && !defined(__clang__) && !defined(strdup)
 +#define strdup _strdup
 +#endif
```

**Comment by rkooo567** [comment]
Seems like this is still failing. @iycheng do you know the answer for questions ^?

**Comment by mattip** [comment]
I rebased to see if by magic anything changed and boringssl now builds on windows.

**Comment by rkooo567** [comment]
@mattip since we don't support ray ha in windows, the best way forward seems to just not include redis for windows build. Do you think it is possible to take this approach instead?

**Comment by mattip** [comment]
> the best way forward seems to just not include redis for windows build

I see `redis-client` is a dependency of `_raylet`, 'gcs_pub_sub_lib`, `gcs_client_lib`, `gcs_table_storage_test_lib`, `gcs_table_storage_lib`, `chaos_redis_store_client_test`. Will any of that work without redis?

**Comment by mattip** [comment]
Also see #42098

**Comment by stephanie-wang** [comment]
> > the best way forward seems to just not include redis for windows build
> 
> I see `redis-client` is a dependency of `_raylet`, `gcs_pub_sub_lib`, `gcs_client_lib`, `gcs_table_storage_test_lib`, `gcs_table_storage_lib`, `chaos_redis_store_client_test`. Will any of that work without redis?

Is it possible to include redis-client but not redis server?

**Comment by mattip** [comment]
> Is it possible to include redis-client but not redis sever?

I don't think that would solve the current problem. The dependency is [redis_client depends on hiredis](https://github.com/ray-project/ray/blob/18b2cab6ca79220d1d0470a60b91b1eb95c1e4a6/BUILD.bazel#L2113), and hiredis [does not build on windows](https://buildkite.com/ray-project/premerge/builds/15210#018c884b-7576-47ac-bba1-0703a94ae38b/6-11203) due to a failure to build boringssl:
```
ERROR: C:/tmp/wqe5u2zk/external/com_github_redis_hiredis/BUILD.bazel:28:11: Compiling ssl.c failed: (Exit 2): cl.exe failed: error executing command
...
external/boringssl/src/include\openssl/base.h(299): error C2059: syntax error: '<parameter-list>'
external/boringssl/src/include\openssl/pkcs7.h(123): error C2059: syntax error: '<parameter-list>'
external/boringssl/src/include\openssl/x509.h(186): error C2059: syntax error: '<parameter-list>'
external/boringssl/src/include\openssl/x509.h(189): error C2059: syntax error: '<parameter-list>'
external/boringssl/src/include\openssl/x509.h(325): error C2059: syntax error: '('
```

**Comment by rkooo567** [comment]
@mattip everything will work without redis actually. It is only used when GCS HA is enabled.

**Comment by rkooo567** [comment]
is it possible to just make it build mocked version if it is used for Windows?

**Comment by mattip** [comment]
> It is only used when GCS HA

What is the name of GCS HA in bazel terms? In other words: where is this non-windows-only dependency expressed in the bazel build?

**Comment by mattip** [comment]
> is it possible to just make it build mocked version if it is used for Windows?

If `redis_client` is never needed for windows, it would be better to redo `BUILD.bazel` to include it only when building on non-windows. Is there a way to express this in bazel?

**Comment by rkooo567** [comment]
https://github.com/ray-project/ray/blob/eb19da2add8342a4d34fe3970eb14debbd8f829d/src/ray/gcs/gcs_server/gcs_server.cc#L76

I don't think there's a way to express this in bazel build now, but redis is used only when this condition is met (and in Windows, you always go to IN_MEMORY).

**Comment by rkooo567** [comment]
Not sure if there's a good way to just not include redis in this case. I wonder if we don't import any redis related dependency if the platform is windows within code?

**Comment by rkooo567** [comment]
I think that could work, but we need `#ifndef _WIN32` within code where redis client is imported and used

**Comment by rkooo567** [comment]
@mattip ^ Does it make sense? We can also discuss in person to quickly find a solution.

**Comment by mattip** [comment]
> we also need #ifndef _WIN32 within code where redis client is imported

Yes, makes perfect sense. I will try to push that into this PR.

**Comment by mattip** [comment]
I am trying to untangle this. I have excluded redis_client, but that is not enough. I need to exclude the actual redis sources [from `gcs`](https://github.com/ray-project/ray/blob/d9bd752766a55b4f72ad6a7fab53a948d3cef639/BUILD.bazel#L2264):
```
diff --git a/BUILD.bazel b/BUILD.bazel
index c1453b0037..f575eac8d0 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -2268,13 +2285,24 @@ ray_cc_library(
         ],
         exclude = [
             "src/ray/gcs/*_test.cc",
+            # These are conditionally included in ray_client
+            "src/ray/gcs/asio.cc",
+            "src/ray/gcs/redis_async_context.cc",
+            "src/ray/gcs/redis_client.cc",
+            "src/ray/gcs/redis_context.cc",
         ],
     ),
     hdrs = glob([
-        "src/ray/gcs/*.h",
-    ]),
+            "src/ray/gcs/*.h",
+        ],
+        exclude = [
+            "src/ray/gcs/asio.h",
+            "src/ray/gcs/redis_async_context.h",
+            "src/ray/gcs/redis_client.h",
+            "src/ray/gcs/redis_context.h",
+        ],
+    ),
     deps = [
-        ":hiredis",
         ":node_manager_fbs",
         ":node_manager_rpc",
         ":ray_common",
@@ -2284,7 +2312,10 @@ ray_cc_library(
         "//src/ray/protobuf:gcs_service_cc_proto",
         "//src/ray/util",
         "@boost//:asio",
-    ],
+    ] + select({
+        "@bazel_tools//src/conditions:windows": [],
+        "//conditions:default": [":hiredis", ":redis_client"],
+    }),
 )

 ray_cc_test(
```

but then various parts of the gcs table storage do not work. Is there someone who understands the interactions between redis and the gcs table storage who can give me some guidance?

**Comment by rkooo567** [comment]
cc @iycheng can you make some comments on ^?

**Comment by iycheng** [comment]
@mattip I think exclude hiredis building from windows actually needs more work. let me take a look at the building failure first.

**Comment by iycheng** [comment]
> This diff will fix the failure in `alloc.c`. There is still a failure to build boringssl. I don't really understand a few things:
> 
> * what changed, why does master build boringssl and this version does not?
> * why use [boringssl for hiredis](https://github.com/ray-project/ray/blob/65ed62d512720c037f545d52cfc653dc830a4589/bazel/BUILD.hiredis#L48), and [openssl for redis](https://github.com/ray-project/ray/blob/65ed62d512720c037f545d52cfc653dc830a4589/bazel/BUILD.redis#L35)? It seems openssl is more prevalent, especially in the documentation.
> * If we are talking about security, why is openssl [pinned to the very old openssl v1.1.1f](https://github.com/ray-project/ray/blob/65ed62d512720c037f545d52cfc653dc830a4589/bazel/ray_deps_setup.bzl#L252) when there are security releases, the final 1.1.1 version [is 1.1.1w](https://github.com/openssl/openssl/tags)

Sorry, I don't know the answer
I always think the dependence in bazel is a messy thing. Like grpc depends implicitly on boringssl and openssl is required for some other libs. Also it seems super hard to make things compiled successfully in bazel.

**Comment by mattip** [comment]
Getting closer. Now everything compiles, but there is a linking error:
```
hiredis.lib(ssl.obj) : error LNK2019: unresolved external symbol __imp_CertCloseStore referenced in function redisCreateSSLContextWithOptions
hiredis.lib(ssl.obj) : error LNK2019: unresolved external symbol __imp_CertEnumCertificatesInStore referenced in function redisCreateSSLContextWithOptions
hiredis.lib(ssl.obj) : error LNK2019: unresolved external symbol __imp_CertFreeCertificateContext referenced in function redisCreateSSLContextWithOptions
hiredis.lib(ssl.obj) : error LNK2019: unresolved external symbol __imp_CertOpenSystemStoreA referenced in function redisCreateSSLContextWithOptions
```

[stack overflow](https://stackoverflow.com/questions/37522654/linking-with-openssl-lib-statically) thinks the link is missing Windows' crypt32.lib library. This would be the link command for hiredis.

**Comment by mattip** [comment]
DCO run is failing with
> Commit sha: [b31089f](https://github.com/ray-project/ray/pull/40217/commits/b31089f833345f283801a9ad72335ccf2116bf47), Author: Yi Cheng, Committer: Yi Cheng; The sign-off is missing.

**Comment by mattip** [comment]
Ahh, `additional_linker_inputs` is not available in this version of bazel, I changed it to [linkopts](https://docs.bazel.build/versions/5.4.1/be/c-cpp.html#cc_library.linkopts).

**Comment by iycheng** [comment]
all passed. merging it.

**Pull request merged**